### PR TITLE
Refactor user preference service

### PR DESCRIPTION
**Pull request by sqreek**
Major changes:
* Move deprecated endpoint to separate Resource
* Introduce FilteredPreferennce to handle duplicate code
* Refactor tests, by moving all occurence and duplicate code to one Helper class

//cc @Wikia/services-team

### CODE REVIEW

File: service/user-preference/src/main/java/com/wikia/userpreference/helper/AuthHelper.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package com.wikia.userpreference.helper;
+
+import com.wikia.dropwizard.auth.AuthenticatedUser;
+import com.wikia.pandora.http.StandardResponses;
+
+import java.util.Objects;
+
+import javax.ws.rs.WebApplicationException;
+
+public class AuthHelper {
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/helper/AuthHelper.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCities.java [hunk 2]
```diff
@@ -2,20 +2,15 @@
...
 import com.wikia.userpreference.domain.Preference;
 
-import com.google.common.collect.ImmutableMap;
-
-import java.sql.SQLException;
 import java.util.List;
 import java.util.Optional;
 
+@Deprecated
 public interface WikiCities {
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCities.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)
MAJOR Add the missing @deprecated Javadoc tag. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)

File: service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceLegacyResource.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.core.Response;
+
+@Deprecated
+public class UserPreferenceLegacyResource {
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceLegacyResource.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 4]
```diff
@@ -30,21 +31,18 @@
 
 import javax.inject.Inject;
 
+@Deprecated
 public class WikiCitiesDAO implements WikiCities {
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)
MAJOR Add the missing @deprecated Javadoc tag. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 5]
```diff
@@ -7,12 +7,14 @@
 
 import com.wikia.dropwizard.databases.jooq.JooqConfigs;
 import com.wikia.jooq.generated.wikicities.tables.records.UserPropertiesRecord;
+import com.wikia.userpreference.domain.FilteredPreferences;
 import com.wikia.userpreference.domain.Preference;
 import com.wikia.userpreference.domain.Whitelist;
-import com.google.common.collect.ImmutableMap;
 
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: MINOR Remove this unused import 'com.google.common.collect.ImmutableMap'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 9 issues:
* MAJOR 4 major
* MINOR 2 minor
* INFO 3 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Remove this unused import 'java.sql.SQLException'.](https://github.com/Wikia/pandora/blob/945288232d4220eced85f3b25c45ac9c6a064b63/service/user-preference/src/main/java/com/wikia/userpreference/dao/StoredPreferences.java#L5) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCities.java [hunk 6]
```diff
@@ -2,20 +2,17 @@
...
-
-import java.sql.SQLException;
 import java.util.List;
 import java.util.Optional;
-
+/**
+ * @deprecated use StoredPreferences
+ */
+@Deprecated
 public interface WikiCities {
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCities.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceLegacyResource.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            type = "apiKey"),
+        @Authorization(
+            value = SwaggerBundleProvider.EXTERNAL_AUTH_NAME,
+            type = "apiKey")
+    },
+    produces = MediaType.APPLICATION_JSON,
+    description = "User preference resources.")
+@Path("/")
+@Produces(MediaType.APPLICATION_JSON)
+public class UserPreferenceLegacyResource {
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceLegacyResource.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java [hunk 8]
```diff
@@ -30,21 +29,21 @@
 
 import javax.inject.Inject;
 
+/**
+ * @deprecated use StoredPreferences
+ */
+@Deprecated
 public class WikiCitiesDAO implements WikiCities {
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues:
* INFO 3 info

Watch the comments in this conversation to review them.

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [1 duplicated blocks of code must be removed.](https://github.com/Wikia/pandora/blob/9ecfa9209eebe80141971cb32e4f311b1fbdabfe/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceLegacyResourceIntegrationTest.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=common-java%3ADuplicatedBlocks)
* MAJOR [1 duplicated blocks of code must be removed.](https://github.com/Wikia/pandora/blob/9ecfa9209eebe80141971cb32e4f311b1fbdabfe/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceIntegrationTest.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=common-java%3ADuplicatedBlocks)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Remove this unused import 'javax.ws.rs.core.MediaType'.](https://github.com/Wikia/pandora/blob/8d63fc0989234081ac8c91ffea1a2b2795136cd3/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceIntegrationTest.java#L33) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by harnash** [comment]
**tl;dr:** LGTM

After reviewing the original PR and all comments regarding code which was moved to this PR we decided to fix minor bugs:

```
if (filteredPreferences.getWhiteListedPreferences().isEmpty() &&
        !filteredPreferences.getInvalidPreferenceNames().isEmpty()) {
      return true;

change to: 

-> isValidListSomething()

change 
createInsertUserPropertiesStatement -> createInsertUserPropertiesStatement(database.dslForWrites(DB_NAME));


change to copy array
whiteListedPreferences = preferences.stream().collect(Collectors.toList());
```
in separate PR to allow merging this branch and lower the technical dept now.

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCities.java (hunk 6 above)
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCities.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceLegacyResource.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
```
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/resources/UserPreferenceLegacyResource.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java (hunk 8 above)
Comment (WikiaTech) [inline service/user-preference/src/main/java/com/wikia/userpreference/dao/WikiCitiesDAO.java]: INFO Do not forget to remove this deprecated code someday. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 2 major
* INFO 3 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [1 duplicated blocks of code must be removed.](https://github.com/Wikia/pandora/blob/d7ecd143317178b88d8f47712ea58b68c338d870/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceLegacyResourceIntegrationTest.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=common-java%3ADuplicatedBlocks)
* MAJOR [1 duplicated blocks of code must be removed.](https://github.com/Wikia/pandora/blob/d7ecd143317178b88d8f47712ea58b68c338d870/service/user-preference/src/test/java/com/wikia/userpreference/resources/UserPreferenceResourceIntegrationTest.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=common-java%3ADuplicatedBlocks)
